R/linear_pool_sample.R

```diff
@@ -42,15 +35,77 @@ linear_pool_sample <- function(model_out_tbl, weights = NULL,
     )
     weights_col_name <- "weight"
   }
+  weight_by_cols <- colnames(weights)[colnames(weights) != weights_col_name]
   unique_weights <- unique(weights[[weights_col_name]])
 
-  if (length(unique_weights) != 1 || length(unique_provided_samples) != 1 || !is.null(n_output_samples)) {
-    cli::cli_abort(
-      "The requested ensemble calculation doesn't satisfy all conditions:
-      1) {.arg model_out_tbl} contains the same number of samples from each component model,
-      2) {.arg weights} are {.val NULL} or equal for every model,
-      3) {.arg n_output_samples} = {.val NULL}"
-    )
+  if (length(unique_weights) != 1) {
+    cli::cli_abort("{.arg weights} must be {.val NULL} or equal for every model")
+  }
+
+  if (!is.null(n_output_samples)) {
+    samples_per_combo <- model_out_tbl |>
+      dplyr::group_by(dplyr::across(dplyr::all_of(c("model_id", compound_taskid_set)))) |>
+      dplyr::summarize(provided_samples = length(unique(.data[["output_type_id"]]))) |>
+      dplyr::ungroup() |>
+      tidyr::complete(
+        !!!rlang::syms(c("model_id", compound_taskid_set)),
+        fill = list(provided_samples = 0)
+      ) |>
+      dplyr::left_join(weights, weight_by_cols) |>
+      dplyr::mutate(target_samples = ifelse(
+        .data[["provided_samples"]] <= 0,
+        0,
+        floor(.data[[weights_col_name]] * n_output_samples)
+      ))
+
+    if (!is.null(compound_taskid_set)) {
+      samples_per_combo <- split(samples_per_combo, f = samples_per_combo[, compound_taskid_set])
+    } else {
+      samples_per_combo <- list(samples_per_combo)
+    }
+    # deal with n_output_samples not divisible evenly among component models
+    samples_per_combo <- samples_per_combo |>
+      purrr::map(.f = function(split_per_combo) {
+        valid_models <- split_per_combo$model_id[split_per_combo$provided_samples > 0]
+        split_per_combo[[weights_col_name]] <-
+          ifelse(split_per_combo$provided_samples == 0, 0, 1 / length(valid_models))
+        split_per_combo$target_samples <-
+          ifelse(split_per_combo$target_samples == 0, 0, floor(split_per_combo[[weights_col_name]] * n_output_samples))
+
+        actual_output_samples <- split_per_combo |>
+          dplyr::group_by(dplyr::across(dplyr::all_of(compound_taskid_set))) |>
+          dplyr::summarize(output_samples = sum(.data[["target_samples"]])) |>
+          dplyr::pull(.data[["output_samples"]]) |>
+          unique()
```

Comment:
Since we are operating on `split_per_combo`, which was split based on the `compound_taskid_set`, I think this simpler computation is available:
```suggestion
        actual_output_samples <- sum(split_per_combo$target_samples)
```